File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03148/p03148_num4_parsed.py
from collections import defaultdict

# Reading input
N, K = map(int, input().split())
sushi = []
for _ in range(N):
    t, d = map(int, input().split())
    sushi.append((t, d))

# Sorting sushi based on deliciousness in descending order
sushi.sort(key=lambda x: -x[1])

# Creating a dictionary to keep track of the count of each topping
topping_count = defaultdict(int)

# List to store the deliciousness of eaten sushi
eaten_sushi_deliciousness = []

# Loop to find the maximum satisfaction
base_total_deliciousness = 0
variety_bonus = 0
for i in range(K):
    t, d = sushi[i]
    base_total_deliciousness += d
    topping_count[t] += 1
    eaten_sushi_deliciousness.append(d)

# Calculating the variety bonus
variety_bonus = sum([topping_count[t] ** 2 for t in topping_count])

# Calculating maximum satisfaction
max_satisfaction = base_total_deliciousness + variety_bonus

# Printing the maximum satisfaction
print(max_satisfaction)

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03148/p03148_num4_parsed.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
using namespace std;

bool compare(pair<int, int> p1, pair<int, int> p2)
{
    return p1.second > p2.second;
}

int main()
{
    int N, K;
    cin >> N >> K;

    vector<pair<int, int>> sushi(N);
    for(int i=0; i<N; i++)
        cin >> sushi[i].first >> sushi[i].second;
    
    sort(sushi.begin(), sushi.end(), compare);
    
    unordered_map<int, int> count;
    long long baseTotalDeliciousness = 0;
    long long varietyBonus = 0;
    long long maxSatisfaction = 0;
    int curIndex = 0;
    
    for(int i=0; i<K; i++)
    {
        baseTotalDeliciousness += sushi[i].second;
        if(count[sushi[i].first] == 0)
        {
            varietyBonus++;
            count[sushi[i].first]++;
        }
        else
            count[sushi[i].first]++;
    }
    
    maxSatisfaction = baseTotalDeliciousness + varietyBonus * varietyBonus;
    
    for(int i=K; i<N; i++)
    {
        if(count[sushi[i].first] == 0)
        {
            varietyBonus++;
            count[sushi[i].first]++;
            
            while(count[sushi[curIndex].first] > 1)
            {
                count[sushi[curIndex].first]--;
                curIndex++;
            }
            
            baseTotalDeliciousness -= sushi[curIndex].second;
            curIndex++;
        }
        
        long long satisfaction = baseTotalDeliciousness + varietyBonus * varietyBonus;
        maxSatisfaction = max(maxSatisfaction, satisfaction);
    }
    
    cout << maxSatisfaction << endl;

    return 0;
}